src/main/java/com/volmit/iris/core/TreeManager.java

```diff
@@ -0,0 +1,308 @@
+
+        // Must be iris world
+        if (!IrisToolbelt.isIrisWorld(event.getWorld())) {
+            Iris.debug(this.getClass().getName() + " passed it off to vanilla since not an Iris world");
+            return;
+        }
+
+        // Get world access
+        IrisAccess worldAccess = IrisToolbelt.access(event.getWorld());
+        if (worldAccess == null) {
+            Iris.debug(this.getClass().getName() + " passed it off to vanilla because could not get IrisAccess for this world");
+            Iris.reportError(new NullPointerException(event.getWorld().getName() + " could not be accessed despite being an Iris world"));
+            return;
+        }
+
+        // Return null if not enabled
+        if (!worldAccess.getCompound().getRootDimension().getTreeSettings().isEnabled()) {
+            Iris.debug(this.getClass().getName() + " cancelled because tree overrides are disabled");
+            return;
+        }
+
+        Iris.debug("Sapling grew @ " + event.getLocation() + " for " + event.getSpecies().name() + " usedBoneMeal is " + event.isFromBonemeal());
+
+        // Calculate size, type & placement
+        // TODO: Patch algorithm to retrieve saplings, as it's not working as intended (only ever returns 1x1)
+        Cuboid saplingPlane = getSaplings(event.getLocation(), blockData -> event.getLocation().getBlock().getBlockData().equals(blockData), event.getWorld());
+        Iris.debug("Sapling plane is: " + saplingPlane.getSizeX() + " by " + saplingPlane.getSizeZ());
+
+        // Find best object placement based on sizes
+        IrisObjectPlacement placement = getObjectPlacement(worldAccess, event.getLocation(), event.getSpecies(), new IrisTreeSize(1, 1));
+
+        // If none were found, just exit
+        if (placement == null) {
+            Iris.debug(this.getClass().getName() + " had options but did not manage to find objectPlacements for them");
+            return;
+        }
+
+        // Delete existing saplings
+        saplingPlane.forEach(block -> block.setType(Material.AIR));
+
+        // Get object from placer
+        IrisObject object = worldAccess.getData().getObjectLoader().load(placement.getPlace().getRandom(RNG.r));
+
+        // Create object placer
+        IObjectPlacer placer = new IObjectPlacer() {
+
+            @Override
+            public int getHighest(int x, int z) {
+                return event.getWorld().getHighestBlockYAt(x, z);
+            }
+
+            @Override
+            public int getHighest(int x, int z, boolean ignoreFluid) {
+                return event.getWorld().getHighestBlockYAt(x, z, ignoreFluid ? HeightMap.OCEAN_FLOOR : HeightMap.WORLD_SURFACE);
+            }
+
+            @Override
+            public void set(int x, int y, int z, BlockData d) {
+                Block b = event.getWorld().getBlockAt(x, y, z);
+
+                // Set the block
+                b.setBlockData(d, false);
```

Comment:
You are still setting the block here, and you are adding the state to the list of cancelled events.

```java
KList<BlockState> newStates = new KList<>();

@Override
public void set(int x, int y, int z, BlockData d) {
    Block b = event.getWorld().getBlockAt(x, y, z);
    BlockState state = b.getState();
    state.setBlockData(d); // Change the block data in the state
    newStates.add(state); // Add the PENDING state to this list.
    // The block will not actually change unless you use state.update()
}
```

When the new event is fired and NOT cancelled, you can place all the blocks later by updating the states